extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/context/TransactionContext.java

```diff
@@ -105,12 +107,25 @@ public ContextState getState() {
         if (instanceHandle != null) {
             return instanceHandle.get();
         } else if (creationalContext != null) {
+            Transaction currentTransaction = getCurrentTransaction();
             T createdInstance = contextual.create(creationalContext);
             instanceHandle = new ContextInstanceHandleImpl<>((InjectableBean<T>) contextual, createdInstance,
                     creationalContext);
 
             contextState.put(contextual, instanceHandle);
 
+            synchronized (transactions) {
```

Comment:
after making the `transactions` Map a `ConcurrentMap`, the synchronized can be removed; the following operations can be made atomic with a little bit of care, e.g. using `computeIfAbsent` instead of `get`.
Even better, for optimal performance, to a `get` first, and then resort to `computeIfAbsent` only if the first optimistic get operation failed.